test/externalTests/plugins/testCommon.js

```diff
@@ -77,136 +78,124 @@ function inject (bot) {
+    function dig (position) {
+      return bot.dig(bot.blockAt(position))
     }
 
-    function placeAndResume (position, item) {
+    async function place (position, item) {
       // console.log('place and resume with', item)
-      setInventorySlot(36, new Item(item.type, 1, 0), () => {
-        placeBlock(36, position, resume)
-      })
-    }
-
-    function resume () {
-      resetBlocksToSuperflat(cb)
+      await setInventorySlot(36, new Item(item.type, 1, 0))
+      await placeBlock(36, position)
     }
   }
 
-  function placeBlock (slot, position, cb) {
+  async function placeBlock (slot, position) {
     bot.setQuickBarSlot(slot - 36)
     // always place the block on the top of the block below it, i guess.
     const referenceBlock = bot.blockAt(position.plus(new Vec3(0, -1, 0)))
-    bot.placeBlock(referenceBlock, new Vec3(0, 1, 0), cb)
+    return bot.placeBlock(referenceBlock, new Vec3(0, 1, 0))
   }
 
   // always leaves you in creative mode
-  function resetState (cb) {
-    callbackChain([
-      becomeCreative,
-      clearInventory,
-      (cb) => {
-        // console.log('start flying')
-        bot.creative.startFlying()
-        teleport(new Vec3(0, 4, 0), cb)
-      },
-      cb => bot.waitForChunksToLoad(cb),
-      resetBlocksToSuperflat,
-      (cb) => { setTimeout(cb, 1000) },
-      clearInventory
-    ], cb)
+  async function resetState () {
+    await becomeCreative()
+    await clearInventory()
+    bot.creative.startFlying()
+    await teleport(new Vec3(0, 4, 0))
+    await bot.waitForChunksToLoad()
+    await resetBlocksToSuperflat()
+    await sleep(1000)
+    await clearInventory()
   }
 
-  function becomeCreative (cb) {
+  async function becomeCreative () {
     // console.log('become creative')
-    setCreativeMode(true, cb)
+    return setCreativeMode(true)
   }
 
-  function becomeSurvival (cb) {
-    setCreativeMode(false, cb)
+  async function becomeSurvival () {
+    return setCreativeMode(false)
   }
 
-  function setCreativeMode (value, cb) {
-    // this function behaves the same whether we start in creative mode or not.
-    // also, creative mode is always allowed for ops, even if server.properties says force-gamemode=true in survival mode.
-
-    function onMessage (jsonMsg) {
-      // console.log(jsonMsg)
-      switch (jsonMsg.translate) {
-        case 'commands.gamemode.success.self':
-        case 'gameMode.changed':
-          // good.
-          bot.removeListener('message', onMessage)
-          clearTimeout(timeOut)
-          return cb()
-        case 'commands.generic.permission':
-          sayEverywhere('ERROR: I need to be an op (allow cheats).')
-          bot.removeListener('message', onMessage)
-        // at this point we just wait forever.
-        // the intention is that someone ops us while we're sitting here, then you kill and restart the test.
+  async function setCreativeMode (value) {
+    return new Promise((resolve) => {
```

Comment:
rather than having new Promise, use something like this:

```js
while (true) {
  const { translate } = once(bot, 'message')
  if (translate == 'gameMode.changed') {
    // ...
    return
  }
}
```

also, to timeout, simply use `Promise.race`

Reply:
You can import `once` from events

Reply:
... Why would you do that though? That looks way worse

Reply:
OK, I changed my mind after thinking about it. I don't think Promise.race is a good solution though, since you cannot cancel the event listeners. Will probably make a helper function of some sort